File: src/market_data/store.rs
use std::collections::HashMap;

use anyhow::Result;
use chrono::NaiveDate;

use super::{AssetId, AssetRegistryEntry, FxRateKind, FxRatePoint, PriceKind, PricePoint};

#[async_trait::async_trait]
pub trait MarketDataStore: Send + Sync {
    async fn get_price(
        &self,
        asset_id: &AssetId,
        date: NaiveDate,
        kind: PriceKind,
    ) -> Result<Option<PricePoint>>;

    /// Get all prices for an asset across all time.
    async fn get_all_prices(&self, asset_id: &AssetId) -> Result<Vec<PricePoint>>;

    async fn put_prices(&self, prices: &[PricePoint]) -> Result<()>;

    async fn get_fx_rate(
        &self,
        base: &str,
        quote: &str,
        date: NaiveDate,
        kind: FxRateKind,
    ) -> Result<Option<FxRatePoint>>;

    /// Get all FX rates for a currency pair across all time.
    async fn get_all_fx_rates(&self, base: &str, quote: &str) -> Result<Vec<FxRatePoint>>;

    async fn put_fx_rates(&self, rates: &[FxRatePoint]) -> Result<()>;

    async fn get_asset_entry(&self, asset_id: &AssetId) -> Result<Option<AssetRegistryEntry>>;

    async fn upsert_asset_entry(&self, entry: &AssetRegistryEntry) -> Result<()>;
}

pub struct NullMarketDataStore;

#[async_trait::async_trait]
impl MarketDataStore for NullMarketDataStore {
    async fn get_price(
        &self,
        _asset_id: &AssetId,
        _date: NaiveDate,
        _kind: PriceKind,
    ) -> Result<Option<PricePoint>> {
        Ok(None)
    }

    async fn get_all_prices(&self, _asset_id: &AssetId) -> Result<Vec<PricePoint>> {
        Ok(Vec::new())
    }

    async fn put_prices(&self, _prices: &[PricePoint]) -> Result<()> {
        Ok(())
    }

    async fn get_fx_rate(
        &self,
        _base: &str,
        _quote: &str,
        _date: NaiveDate,
        _kind: FxRateKind,
    ) -> Result<Option<FxRatePoint>> {
        Ok(None)
    }

    async fn get_all_fx_rates(&self, _base: &str, _quote: &str) -> Result<Vec<FxRatePoint>> {
        Ok(Vec::new())
    }

    async fn put_fx_rates(&self, _rates: &[FxRatePoint]) -> Result<()> {
        Ok(())
    }

    async fn get_asset_entry(&self, _asset_id: &AssetId) -> Result<Option<AssetRegistryEntry>> {
        Ok(None)
    }

    async fn upsert_asset_entry(&self, _entry: &AssetRegistryEntry) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct MemoryMarketDataStore {
    prices: tokio::sync::Mutex<HashMap<(AssetId, NaiveDate, PriceKind), PricePoint>>,
    fx_rates: tokio::sync::Mutex<HashMap<(String, String, NaiveDate, FxRateKind), FxRatePoint>>,
    assets: tokio::sync::Mutex<HashMap<AssetId, AssetRegistryEntry>>,
}

impl MemoryMarketDataStore {
    pub fn new() -> Self {
        Self::default()
    }
}

#[async_trait::async_trait]
impl MarketDataStore for MemoryMarketDataStore {
    async fn get_price(
        &self,
        asset_id: &AssetId,
        date: NaiveDate,
        kind: PriceKind,
    ) -> Result<Option<PricePoint>> {
        let prices = self.prices.lock().await;
        Ok(prices.get(&(asset_id.clone(), date, kind)).cloned())
    }

    async fn get_all_prices(&self, asset_id: &AssetId) -> Result<Vec<PricePoint>> {
        let prices = self.prices.lock().await;
        Ok(prices
            .iter()
            .filter(|((id, _, _), _)| id == asset_id)
            .map(|(_, p)| p.clone())
            .collect())
    }

    async fn put_prices(&self, prices: &[PricePoint]) -> Result<()> {
        if prices.is_empty() {
            return Ok(());
        }
        let mut store = self.prices.lock().await;
        for price in prices {
            store.insert(
                (price.asset_id.clone(), price.as_of_date, price.kind),
                price.clone(),
            );
        }
        Ok(())
    }

    async fn get_fx_rate(
        &self,
        base: &str,
        quote: &str,
        date: NaiveDate,
        kind: FxRateKind,
    ) -> Result<Option<FxRatePoint>> {
        let fx_rates = self.fx_rates.lock().await;
        let base = base.trim().to_uppercase();
        let quote = quote.trim().to_uppercase();
        Ok(fx_rates
            .get(&(base, quote, date, kind))
            .cloned())
    }

    async fn get_all_fx_rates(&self, base: &str, quote: &str) -> Result<Vec<FxRatePoint>> {
        let fx_rates = self.fx_rates.lock().await;
        let base = base.trim().to_uppercase();
        let quote = quote.trim().to_uppercase();
        Ok(fx_rates
            .iter()
            .filter(|((b, q, _, _), _)| b == &base && q == &quote)
            .map(|(_, r)| r.clone())
            .collect())
    }

    async fn put_fx_rates(&self, rates: &[FxRatePoint]) -> Result<()> {
        if rates.is_empty() {
            return Ok(());
        }
        let mut store = self.fx_rates.lock().await;
        for rate in rates {
            let normalized = FxRatePoint {
                base: rate.base.trim().to_uppercase(),
                quote: rate.quote.trim().to_uppercase(),
                ..rate.clone()
            };
            store.insert(
                (
                    normalized.base.clone(),
                    normalized.quote.clone(),
                    normalized.as_of_date,
                    normalized.kind,
                ),
                normalized,
            );
        }
        Ok(())
    }

    async fn get_asset_entry(&self, asset_id: &AssetId) -> Result<Option<AssetRegistryEntry>> {
        let assets = self.assets.lock().await;
        Ok(assets.get(asset_id).cloned())
    }

    async fn upsert_asset_entry(&self, entry: &AssetRegistryEntry) -> Result<()> {
        let mut assets = self.assets.lock().await;
        assets.insert(entry.id.clone(), entry.clone());
        Ok(())
    }
}

File: ts/src/market-data/store.ts
/**
 * Market data store interface and implementations.
 *
 * Port of the Rust `market_data::store` module. Defines the async store
 * trait (interface) and provides NullMarketDataStore and MemoryMarketDataStore
 * implementations.
 */

import { AssetId } from './asset-id.js';
import type {
  PriceKind,
  FxRateKind,
  PricePoint,
  FxRatePoint,
  AssetRegistryEntry,
} from './models.js';

// ---------------------------------------------------------------------------
// MarketDataStore interface
// ---------------------------------------------------------------------------

/** Async store for market data: prices, FX rates, and asset registry entries. */
export interface MarketDataStore {
  get_price(assetId: AssetId, date: string, kind: PriceKind): Promise<PricePoint | null>;

  get_all_prices(assetId: AssetId): Promise<PricePoint[]>;

  put_prices(prices: PricePoint[]): Promise<void>;

  get_fx_rate(
    base: string,
    quote: string,
    date: string,
    kind: FxRateKind,
  ): Promise<FxRatePoint | null>;

  get_all_fx_rates(base: string, quote: string): Promise<FxRatePoint[]>;

  put_fx_rates(rates: FxRatePoint[]): Promise<void>;

  get_asset_entry(assetId: AssetId): Promise<AssetRegistryEntry | null>;

  upsert_asset_entry(entry: AssetRegistryEntry): Promise<void>;
}

// ---------------------------------------------------------------------------
// NullMarketDataStore
// ---------------------------------------------------------------------------

/** A no-op store: all getters return null/empty, all puts are no-ops. */
export class NullMarketDataStore implements MarketDataStore {
  async get_price(_assetId: AssetId, _date: string, _kind: PriceKind): Promise<PricePoint | null> {
    return null;
  }

  async get_all_prices(_assetId: AssetId): Promise<PricePoint[]> {
    return [];
  }

  async put_prices(_prices: PricePoint[]): Promise<void> {}

  async get_fx_rate(
    _base: string,
    _quote: string,
    _date: string,
    _kind: FxRateKind,
  ): Promise<FxRatePoint | null> {
    return null;
  }

  async get_all_fx_rates(_base: string, _quote: string): Promise<FxRatePoint[]> {
    return [];
  }

  async put_fx_rates(_rates: FxRatePoint[]): Promise<void> {}

  async get_asset_entry(_assetId: AssetId): Promise<AssetRegistryEntry | null> {
    return null;
  }

  async upsert_asset_entry(_entry: AssetRegistryEntry): Promise<void> {}
}

// ---------------------------------------------------------------------------
// MemoryMarketDataStore
// ---------------------------------------------------------------------------

/**
 * In-memory store backed by Maps.
 *
 * Key schemes:
 * - Prices: `${assetId}|${date}|${kind}`
 * - FX rates: `${BASE}|${QUOTE}|${date}|${kind}` (base/quote normalized to uppercase)
 * - Asset entries: AssetId string
 */
export class MemoryMarketDataStore implements MarketDataStore {
  readonly #prices = new Map<string, PricePoint>();
  readonly #fxRates = new Map<string, FxRatePoint>();
  readonly #assetEntries = new Map<string, AssetRegistryEntry>();

  // -- Prices ---------------------------------------------------------------

  async get_price(assetId: AssetId, date: string, kind: PriceKind): Promise<PricePoint | null> {
    const key = priceKey(assetId.asStr(), date, kind);
    return this.#prices.get(key) ?? null;
  }

  async get_all_prices(assetId: AssetId): Promise<PricePoint[]> {
    const prefix = assetId.asStr();
    const results: PricePoint[] = [];
    for (const [key, point] of this.#prices) {
      if (key.startsWith(prefix + '|')) {
        results.push(point);
      }
    }
    return results;
  }

  async put_prices(prices: PricePoint[]): Promise<void> {
    for (const p of prices) {
      const key = priceKey(p.asset_id.asStr(), p.as_of_date, p.kind);
      this.#prices.set(key, p);
    }
  }

  // -- FX rates -------------------------------------------------------------

  async get_fx_rate(
    base: string,
    quote: string,
    date: string,
    kind: FxRateKind,
  ): Promise<FxRatePoint | null> {
    const key = fxKey(base.toUpperCase(), quote.toUpperCase(), date, kind);
    return this.#fxRates.get(key) ?? null;
  }

  async get_all_fx_rates(base: string, quote: string): Promise<FxRatePoint[]> {
    const prefix = `${base.toUpperCase()}|${quote.toUpperCase()}|`;
    const results: FxRatePoint[] = [];
    for (const [key, point] of this.#fxRates) {
      if (key.startsWith(prefix)) {
        results.push(point);
      }
    }
    return results;
  }

  async put_fx_rates(rates: FxRatePoint[]): Promise<void> {
    for (const r of rates) {
      const normalizedBase = r.base.toUpperCase();
      const normalizedQuote = r.quote.toUpperCase();
      const key = fxKey(normalizedBase, normalizedQuote, r.as_of_date, r.kind);
      // Store with normalized base/quote
      const normalized: FxRatePoint = {
        ...r,
        base: normalizedBase,
        quote: normalizedQuote,
      };
      this.#fxRates.set(key, normalized);
    }
  }

  // -- Asset entries --------------------------------------------------------

  async get_asset_entry(assetId: AssetId): Promise<AssetRegistryEntry | null> {
    return this.#assetEntries.get(assetId.asStr()) ?? null;
  }

  async upsert_asset_entry(entry: AssetRegistryEntry): Promise<void> {
    this.#assetEntries.set(entry.id.asStr(), entry);
  }
}

// ---------------------------------------------------------------------------
// Key helpers
// ---------------------------------------------------------------------------

function priceKey(assetIdStr: string, date: string, kind: PriceKind): string {
  return `${assetIdStr}|${date}|${kind}`;
}

function fxKey(base: string, quote: string, date: string, kind: FxRateKind): string {
  return `${base}|${quote}|${date}|${kind}`;
}
